### DISCUSSION TITLE
Expose next scheduled timer event

### DISCUSSION THREAD

**Discussion by moedi21** [arendst/Tasmota#20148] [Ideas]
I am using Tasmota devices mainly for controlling some lights base on the integrated timers. I would like to integrate those devices more into my Homeassistant setup. There, I would like to show the time, when the next timer event will be executed, but I did not find  a way to retrieve this information from the Tasmota devices.
I can request the timer configuration by MQTT, but the exact time of the next event is not accessible, especially if you use the randomization window. For example. if I have a timer set to switch off the light 15 min after sunrise, with a +/- 10 min time window, I would like to know the exact time when this event will happen.

**Comment by barbudor** [comment]
It's not possible because the next time isn't computed in advance.
Every minute, Tasmota check if it's time for a timer event, not when the next event will occur. When it's time, it is.

Somehow, Tasmota is short seing and live in the present without taking care of the future.
Carpe Diem.

**Reply by moedi21** [reply]
So how does randomization work in this case? Does Tasmote throw a coin every minute to decide if it should trigger the timer?

**Reply by arendst** [reply]
At midnight it calculates random times once a day

**Reply by sfromis** [reply]
Every minute, the pre-calculated random window offset is added to the base timer, when it checks if now is the time. That's also behind a weakness of the timers, that Tasmota is not detecting if a timer was to act while Tasmota was down for whatever reason. Hence, rule examples exist for trying to handle this.
https://tasmota.github.io/docs/Rules/#turn-on-light-before-dawn-and-at-dusk_1

**Reply by barbudor** [reply]
@moedi21 The problem is that any calculation in advance could be made wrong by a configuration change between the moment of the calculation and the effective action.
So the best estimation would be listen to timer configuration messages and compute on reception of those message.

**Reply by moedi21** [reply]
Thanks for all the insights to the internals. I am quite new to Tasmota, and I somehow assumed that a timer has a rule (like sunset + random), but also a property 'next trigger time', which is updated after
 - midnight
 - on rule change
 - on trigger event
and my idea was to retrieve this 'next' property. But it seems this is not possible in the way how timers are implemented right now.
